### PR TITLE
Add template for document generation with 'jsdoc'.

### PR DESCRIPTION
**Pull request by wimrijnders**
In essence, this PR defines the subdirectory `docs` as a `jsdoc` template.

**Benefits:**
- allows the usage of partials, in order to DRY common parts of the html files.
- makes available the `jsdoc`-comments, for addition into the documentation.
- enables extraction of data from the source code. For example, the list of edge endpoints `['arrow', 'bar', 'circle']` can now be extracted from the source and inserted into the documentation on generation.

----
In this initial version, the only file that has been changed is `docs/data/dataset.html`. Here, partials have been added to illustrate how common page elements can be DRY'd.

The template has been set up in such a way, that resource files will be copied and that html files can pass through unchanged if no special template tags (`<?js...?>`) are used. This allows for a gradual transition of the html files to templates.

A `README.md` has been added to subdirectory `docs`, with notes on how to use the `jsdoc`-generated comment data.

----
**Usage:**  `gulp docs`

- The result files are placed in subdirectory `gen/docs/`.

**NOTE:** The release procedure will have to be adjusted by adding this gulp command.
The docs-files will then have to be taken from `gen/docs`.

### CODE REVIEW

File: docs/README.md [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+  The standard way of using this appear to be:
+  ```
+  <?js
+    var data = obj;   // Whatever docData is
+    var self = this;
+  ?>
+  ```
+ 
+  But it also appear to be possible to use the elements of docData directly:
+  ```
```
Comment (mbroad) [inline docs/README.md]: You can get proper syntax highlighting by adding the name to your code block.

ie, 
<pre>
```javascript
var docData =...
```
<pre>

Comment (mbroad) [reply docs/README.md]: json is also a valid key word for syntax highlighting

Comment (wimrijnders) [reply docs/README.md]: yeah, I knew that, it just didn't occur to me to use it here.

```js
 var msg = 'js works as well.';
```

<pre>And now I learnt from you how to display ```js itself!</pre>

Comment (wimrijnders) [reply docs/README.md]: Please do keep on reviewing! And yes try it out, please, if only to see if it works on machines other than mine.

File: docs/README.md [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+- Files ending in `.tmpl` are skipped
+- All non-html files are plain copied
+- html-files *can* contain `<?js ?>` tags, but this is not required
+
+
+## Intention
+
+The `docs` directory is treated as a `jsdoc` template, in which the html-files are the template files.
+This allows for a gradual adaptation of the html-files to templates; unchanged html-files will pass 
+through `jsdoc` unchanged.
```
[the remark this answers is not in the document]
Comment (wimrijnders) [reply docs/README.md]: ok. It should be readable as text as well, of course.

File: gulpfile.js [hunk 3]
```diff
@@ -232,5 +233,27 @@ gulp.task('lint', function () {
...
+
+  // Not sure if this is the best way to handle 'cb'; at least it works.
+  var hasError = false;
+  var onError = function(error) {
+    if (error !== undefined && error !== null) {
+      console.error('Error while running task: ' + error);
+      hasError = true;
+      cb();
+    }
+  }
```
Comment (macleodbroad-wf) [inline gulpfile.js]: missing semi-colon

File: gulpfile.js [hunk 4]
```diff
@@ -232,5 +233,27 @@ gulp.task('lint', function () {
...
+
+  rimraf(__dirname + '/' + targetDir, onError);  // Clean up previous generation
+
+  if (!hasError) {
+    var params = '-c ./jsdoc.json -r -t docs -d ' + targetDir;
+    child_exec('node ./node_modules/jsdoc/jsdoc.js ' + params + ' lib', undefined, cb);
+  };
```
Comment (macleodbroad-wf) [inline gulpfile.js]: unnecessary semi-colon

**Comment by macleodbroad-wf** [comment]
I found that the index.html files did not work as I would have expected them to? Maybe this is a misunderstanding on my part.

File: gulpfile.js (hunk 3 above)
Comment (wimrijnders) [reply gulpfile.js]: ok

File: gulpfile.js (hunk 4 above)
Comment (wimrijnders) [reply gulpfile.js]: Due to this, I linted `gulpfile.js` and to my delight, I discovered that it's actually possible. What do you think, should it officially be added to the linting step?

**Comment by wimrijnders** [comment]
The top index.html has this redirect in it which possibly reconnects to the main site. But all the other, lower-level, index.html's should be just fine.

The top-level index.html can be excused; do you have trouble with the other ones as well?

**Comment by mbroad** [comment]
Maybe I'm misunderstanding how to use this.

If I navigate to dataset.html directly, everything looks fine.

If I open the index.html file, I'm redirected, but file not found.

If I copy and paste the entire file structure into the docs the gh-pages branch, replacing existing files, then the index pages work, but show file listings.

It is entirely possible that this is just user error.

In any case the generated dataset.html looks great.

**Comment by wimrijnders** [comment]
You're not misunderstanding anything; that's how the html files are set up. These pages are made to be integrated into the web site, as in `gh-pages`. So, locally the redirect goes nowhere. Another thing I notice is that the links in the menu's are hardcoded, i.e. they take you to somewhere on `http://visjs.org`.

> show file listings

Perhaps you mean that they *don't* show?

As for the greatness of the generated page, it should be just identical to the original version. Next step is to isolate all the common parts in partials - obvious candidates: the HTML at top and bottom, the menu's. And there are plenty more duplicates of data in the doc's themselves, e.g. font settings in both nodes and edges.

Hoping for some help in converting them....

**Comment by mbroad** [comment]
And will that also mean eventually incorporating documentation on the web page that is generated from some of the jsdoc in the source code itself?

**Comment by wimrijnders** [comment]
Oh yes, definitely. That is in fact the major point of this exercise. The rest is just nice consequences.

**Comment by mbroad** [comment]
Excellent!

**Comment by wimrijnders** [comment]
Since the whole point of using `jsdoc` for document generation is to be able to use the header comments, I've added a proof of concept. Diffing should make everything clear.

**Pull request merged**